File: 54_Spiral_Matrix.py
class Solution:
    def spiralOrder(self, matrix):
        """

        :param matrix: List[List[int]]
        :return: List[int]
        """
        result = []
        if len(matrix) == 0:
            return result
        if len(matrix[0]) == 0:
            return result
        leftedge = 0
        rightedge = len(matrix[0]) - 1
        upedge = 0
        downedge = len(matrix) - 1

        while True:
            # 在最顶一行
            for j in range(leftedge, rightedge + 1, 1):
                result.append(matrix[upedge][j])
            upedge += 1
            if upedge > downedge:
                break

            # 在最右边一列
            for i in range(upedge, downedge + 1, 1):
                result.append(matrix[i][rightedge])
            rightedge -= 1
            if rightedge < leftedge:
                break

            # 在最底一行
            for j in range(rightedge, leftedge - 1, -1):
                result.append(matrix[downedge][j])
            downedge -= 1
            if downedge < upedge:
                break

            # 在最左边一列
            for i in range(downedge, upedge - 1, -1):
                result.append(matrix[i][leftedge])
            leftedge += 1
            if leftedge > rightedge:
                break

        return result

A = Solution()
x = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
print(A.spiralOrder(x))

File: 54_Spiral_Matrix.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	vector<int> spiralOrder(vector<vector<int> >& matrix) {
		static const int UP = 0;
		static const int DOWN = 1;
		static const int RIGHT = 2;
		static const int LEFT = 3;
		int direction = RIGHT;
		vector<int> result;
		if(matrix.size() <= 0 || matrix[0].size() <= 0) return result;
		int* tag = new int[matrix.size() * matrix[0].size()];	// tag[i * matrix[0].size() + j] 表示第i行j列
		for(int i = 0; i < matrix.size() * matrix[0].size(); i++) { // Initialize
			tag[i] = 0;
		}
		int i = 0, j = 0;	//当前位置
		while(1) {
			if(i < 0 || i >= matrix.size() || j < 0 || j >= matrix[0].size()) break;
			if(1 == tag[i * matrix[0].size() + j]) break;
			//处理
			result.push_back(matrix[i][j]);
			tag[i * matrix[0].size() + j] = 1;
			// 进一
			switch(direction) {
				case UP:
					if(0 == i || 1 == tag[(i - 1) * matrix[0].size() + j]) {
						direction = RIGHT;
						j++;
					}
					else {
						i--;
					}
					break;
				case DOWN:
					if(matrix.size() - 1 == i || 1 == tag[(i + 1) * matrix[0].size() + j]) {
						direction = LEFT;
						j--;
					}
					else {
						i++;
					}
					break;
				case LEFT:
					if(0 == j || 1 == tag[i * matrix[0].size() + j - 1]) {
						direction = UP;
						i--;
					}
					else {
						j--;
					}
					break;
				case RIGHT:
					if(matrix[0].size() - 1 == j || 1 == tag[i * matrix[0].size() + j + 1]) {
						direction = DOWN;
						i++;
					}
					else {
						j++;
					}
					break;
				default: break;
			}
		}
		delete [] tag;
		return result;
    	}
};

int main() {
	Solution* A = new Solution();
	vector<vector<int> > matrix;
	vector<int> tmp;
	for(int i = 0; i < 3; i++) {
		tmp.clear();
		for(int j = 1; j <= 3; j++) {
			tmp.push_back(i * 3 + j);
		}
		matrix.push_back(tmp);
	}

	cout << "matrix = "<< endl;
	for(int i = 0; i < matrix.size(); i++) {
		for(int j = 0; j < matrix[i].size(); j++) {
			cout << matrix[i][j] << " ";
		}
		cout << endl;
	}

	cout << "result = " << endl;
	vector <int> result;
	result = A->spiralOrder(matrix);
	for(int i = 0; i < result.size(); i++) {
		cout << result[i] << " ";
	}
	cout << endl;
	cout << "result.size() = " << result.size() << endl;

	return 0;

}